docs/proposals/koordlet/20221018-koordlet-support-cgroups-v2.md

```diff
@@ -0,0 +1,324 @@
+
+type Validator interface {
+	Validate(value *int64) error
+}
+```
+
+CgroupFile Implementations:
+
+```go
+var (
+    // case 1: same file name with different content
+    CPUStat   = CgroupFile{ResourceFileName: "cpu.stat", Subfs: "cpu", IsAnolisOS: false}
+    CPUStatV2 = CgroupFile{ResourceFileName: "cpu.stat", Subfs: "", IsAnolisOS: false} // v2 `cpu.stat` includes the contents of v1 `cpuacct.stat` and `cpu.stat`
+    // case 2: different file name with similar functionality
+    CPUShares   = CgroupFile{ResourceFileName: "cpu.shares", Subfs: CgroupCPUDir, IsAnolisOS: false}
+    CPUWeightV2 = CgroupFile{ResourceFileName: "cpu.weight", Subfs: "", IsAnolisOS: false}
+    // case 3: same file name with different os requirement
+    MemoryMin   = CgroupFile{ResourceFileName: "memory.min", Subfs: "memory", IsAnolisOS: true}
+    MemoryMinV2 = CgroupFile{ResourceFileName: "memory.min", Subfs: "", IsAnolisOS: false}
+)
+```
+
+#### Resource Executor
+
+**ResourceExecutor** defines an executable cgroup resource.
+
+```go
+type ResourceExecutor interface {
+	Owner() *OwnerRef // container, pod, qos
+	Key() string
+}
+
+type ReadFunc func(resource ResourceExecutor) (string, error)
+
+type UpdateFunc func(resource ResourceExecutor, value string) error
+
+type MergeFunc func(resource ResourceExecutor, value string) (bool, error)
+```
+
+#### Update
+
+**ResourceUpdater** defines an updatable cgroup resource.
+
+```go
+type ResourceUpdater interface {
+    ResourceExecutor
+	Value() string
+	Update() error
+	Clone() ResourceUpdater
+	Merge() (bool, error) // return false if no need to merge
+}
+```
+
+ResourceUpdater Implementations:
+
+```go
+// CgroupResourceUpdater implements the ResourceUpdater for updating any cgroup file.
+type CgroupResourceUpdater struct {
+	owner               *OwnerRef
+	value               string
+	parentDir           string
+	file                system.CgroupFile
+	updateFunc          UpdateFunc
+	lastUpdateTimestamp time.Time // for cacheable update
+	mergeFunc           MergeFunc // optional; for merge update
+}
+
+func (c *CgroupResourceUpdater) Owner() *OwnerRef {
+    return c.owner
+}
+
+func (c *CgroupResourceUpdater) Key() string {
+    return sysutil.GetCgroupFilePath(c.parentDir, c.file)
+}
+
+func (c *CgroupResourceUpdater) Update() error {
+    return c.updateFunc(c, c.value)
+}
+
+// CommonResourceUpdater defines the general ResourceUpdater implementation, can be used for resctrl, MPAM.
+type CommonResourceUpdater struct {
+	owner               *OwnerRef
+	value               string
+	key                 string
+	file                string
+	updateFunc          UpdateFunc
+	lastUpdateTimestamp time.Time // optional; for cacheable update
+}
+
+// GuestCgroupResourceUpdater implements the ResourceUpdater for cgroup files in guest OS (e.g. for kata-containers).
+type GuestCgroupResourceUpdater struct {
+    *CgroupResourceUpdater
+    sandboxID string // for execute the file operation inside the sandbox
+}
+```
+
```

Comment:
executor and updater are confused, can you explain the relationship between them?

Reply:
The old `Executor` interface is removed.